services/processing/processing.py

```diff
@@ -0,0 +1,117 @@
+import logging
+from typing import Callable, NotRequired, TypedDict
+
+import sentry_sdk
+from celery.exceptions import CeleryError
+from shared.yaml import UserYaml
+from sqlalchemy.orm import Session as DbSession
+
+from database.models.core import Commit
+from database.models.reports import Upload
+from helpers.reports import delete_archive_setting
+from services.archive import ArchiveService
+from services.processing.intermediate import save_intermediate_report
+from services.processing.state import ProcessingState
+from services.report import ProcessingError, RawReportInfo, Report, ReportService
+from services.report.parser.types import VersionOneParsedRawReport
+
+log = logging.getLogger(__name__)
+
+
+class UploadArguments(TypedDict):
+    # TODO(swatinem): migrate this over to `upload_id`
+    upload_pk: int
+
+
+class ProcessingResult(TypedDict):
+    arguments: UploadArguments
+    successful: bool
+    error: NotRequired[dict]
+
+
+@sentry_sdk.trace
+def process_upload(
+    on_processing_error: Callable[[ProcessingError], None],
+    db_session: DbSession,
+    repo_id: int,
+    commit_sha: str,
+    commit_yaml: UserYaml,
+    arguments: UploadArguments,
+) -> dict:
+    upload_id = arguments["upload_pk"]
+
+    commit = (
+        db_session.query(Commit)
+        .filter(Commit.repoid == repo_id, Commit.commitid == commit_sha)
+        .first()
+    )
+    assert commit
+
+    upload = db_session.query(Upload).filter_by(id_=upload_id).first()
+    assert upload
+
+    state = ProcessingState(repo_id, commit_sha)
+    # this in a noop in normal cases, but relevant for task retries:
+    state.mark_uploads_as_processing([upload_id])
+
+    report_service = ReportService(commit_yaml)
+    archive_service = report_service.get_archive_service(commit.repository)
+
+    result = ProcessingResult(arguments=arguments, successful=False)
+
+    try:
+        report = Report()
+        report_info = RawReportInfo()
+        processing_result = report_service.build_report_from_raw_content(
+            report, report_info, upload
+        )
+
+        if error := processing_result.error:
+            on_processing_error(error)  # NOTE: this might throw a `Retry`
+            result["error"] = error.as_dict()
+        else:
+            result["successful"] = True
+        log.info("Finished processing upload", extra={"result": result})
+
+        report_service.update_upload_with_processing_result(upload, processing_result)
+        save_intermediate_report(archive_service, commit_sha, upload_id, report)
+        state.mark_upload_as_processed(upload_id)
+
+        rewrite_or_delete_upload(archive_service, commit_yaml, report_info)
+
+    except CeleryError:
+        raise
```

Comment:
[nit, feel free to ignore] only because I've spent quite some time looking at `ruff` output, one of the rules say you are supposed to re-raise errors like so:

```
except CeleryError as err:
    raise Exception("panic") from err
```
    
The only real difference is that it sets the 1st exception as the direct cause of the re-raise (which is true). Idk if it makes difference in Sentry issues page or logs... Read more: https://stackoverflow.com/questions/24752395/python-raise-from-usage

Reply:
This is an interesting concept. Though I want to explicitly pass on the exception un-chained, as I’m not sure if celery can deal with chained exceptions. And I’m also not sure if adding any exception context in such a way is really provides any value in the first place.